### ISSUE TITLE
Gripper joint states: approach explanation

### ISSUE THREAD

**Issue by JanJericevic** [DFKI-NI/mobipick#16]
Hello,

I'm a master's student currently working on a similar setup as your Mobipick robot - MiR100 base with the UR5e robot arm and a Robotiq 2f85 gripper.
I'm currently working on a unified UR5e & Robotiq 2f85 robot description. I have successfully created the unified xacro that I can use in simulation however, I'm having trouble with gripper joint states when using the real robot and was hoping for your help. 

I'm using the [Universal Robots ROS driver](https://github.com/UniversalRobots/Universal_Robots_ROS_Driver) and this [Robotiq gripper package](https://github.com/TAMS-Group/robotiq).

My gripper is connected directly to the robot flange and I can successfully communicate with the gripper over RTU using `Robotiq2FGripperRtuNode.py)`  from the `robotiq_2f_gripper_control` package. But no node publishes the gripper joint states, only gripper status. Since there are no gripper joint states the tf transforms between the robot and the gripper don't exist when launching the `ur_robot_driver`.

I saw that you have a [gripper joint state publisher](https://github.com/DFKI-NI/mobipick/blob/noetic/mobipick_bringup/nodes/gripper_joint_state_publisher.py) but it looks like it only publishes the state of the joint that is controlled? By that I mean the joint we control to open/close the gripper. Is advertising that one joint enough to get the relative position of all other gripper joints? Or am I missing something?
I'm also a bit confused why you use the `joint_state_publisher` [here](https://github.com/DFKI-NI/mobipick/blob/a718c50a5189e44c28e7438ea6ff6b172de4d802/mobipick_bringup/launch/mobipick_bringup.launch#L82) in the bringup.

Any other advice on how to approach gripper joint states is greatly appreciated.

Best regards,
Jan

**Comment by mintar** [comment]
We had to modify the robotiq URDF a bit since it was broken. In our modified version in this package, we use `mimic` tags for all dependent joints; in other words, we specify that each dependent joint either mirrors the controlled joint or is the exact opposite of the controlled joint (factor -1). Just search for the word "mimic" in the URDFs in this repo.

To get the `mimic` joints to work, we need to run that `joint_state_publisher` that you pointed to. It gets a list of joint state topics as input, parses the URDF for mimic joints, and adds those mimic joint states to the output topic (along with all the input joint states). This is why the `gripper_joint_state_publisher` only needs to publish the controlled joint.

See: http://wiki.ros.org/joint_state_publisher#Dependent_Joints

In Gazebo, this works a little bit differently. Unfortunately, Gazebo ignores the `mimic` tags in the URDF. Therefore, we had to use the `mimic_joint_plugin` for this purpose:

https://github.com/DFKI-NI/mobipick/blob/a718c50a5189e44c28e7438ea6ff6b172de4d802/mobipick_description/urdf/robotiq_2f_140/robotiq_2f_140.gazebo.xacro#L4-L21

In another place, we use SDF to close the kinematic loop, so we don't need the `mimic_joint_plugin`; all joint states are properly simulated using physics:

https://github.com/DFKI-NI/mobipick/blob/a718c50a5189e44c28e7438ea6ff6b172de4d802/mobipick_description/urdf/robotiq_2f_140/robotiq_arg2f_140_model_macro.xacro#L173-L185

The trick here was that in Gazebo SDF, kinematic loops are allowed, while in URDF they are forbidden. So we couldn't use a regular URDF joint to close the loop, but a SDF joint works.

In the future, I would really love to get rid of the `mimic_joint_plugin`, because it's causing all kinds of problems. I just couldn't figure out how to do this using onboard Gazebo means, like SDF joints or transmissions or something.
